### ISSUE TITLE
Feature proposal - shuffle DNS response returned by getaddrinfo() before establishing connection

### ISSUE THREAD

**Issue by mcslr** [urllib3/urllib3#1404]
Hi folks,

I was trying to find if the issue was already discussed and I could not find anything so here it goes.

What do you think about adding feature allowing to shuffle list of DNS responses returned by socket.getaddrino before trying to establish connection.

Similar solution was recently added to curl:
[https://github.com/curl/curl/pull/1694](https://github.com/curl/curl/pull/1694
)

Thoughts?

**Comment by SethMichaelLarson** [comment]
This feature would affect `urllib3/utils/connection.create_connection()` which is basically as low as you can go when it comes to API so I'm curious how we'd make this configurable from any high-level interface.

**Comment by jvanasco** [comment]
this feature would underscore the importance of making a debug object #1071 to cache the upstream IP address.

**Comment by sigmavirus24** [comment]
@jvanasco note that there's no actual decision to do this. In fact, there's no good way to provide a way of doing this.

**Comment by jvanasco** [comment]
@sigmavirus24 i know the first part, did not know the second. i really want to move forward with some agreement on a debug object so i can start a PR for #1071 though ;)

**Comment by jvanasco** [comment]
I'm sorry you thought this was a bad faith intent - it was intended as a a legitimate cross reference against 1071 as a potentially blocking related issue.  The eventual concept of a "Debug" object in 1071 was coalescing my original request to track the IP address and glyph's need to access the active certificate.  Early feedback in that ticket (and some related ones in other projects), and existing workarounds for some people who share these needs, is that this type of data can be somewhat calculated with an secondary request [which assumed getaddrino returns the same info on every invocation, something that is not guaranteed]. If this requested feature were to be implemented and the DNS randomized, it would break the existing behavior workarounds disclosed in 1071 - making it impossible to tell where a request was served from.

**Comment by bhagwatvyas** [comment]
Hi folks,

Are there any plans on adding this support?

**Comment by sethmlarson** [comment]
@josephholsten contacted me about getting a solution to this and being willing to provide a patch. I'm on board for having the functionality. It'd be nice to have a shuffle that guarantees the first two addresses tried are different if there's both an IPv4 and IPv6 address.

The problem in my eyes in how to handle passing to `create_connection()` that the DNS results should be shuffled. A thought I had was to somehow use `socket_options` with a sentinel value (`SHUFFLE_DNS_ADDRESSES`?) that gets filtered out by `create_connection()` instead of called in `setsockopt()`? What do others think of this method? cc @pquentin @sigmavirus24

**Comment by pquentin** [comment]
@sethmlarson I think I'd prefer using another keyword argument rather than piggybacking on `socket_options`, but I'd be happy to add the feature! 👍 

We can probably just use a simple `random.shuffle()` if the shuffle was asked instead of just iterating over the `getaddrinfo()` results. This seems like a two-line change, but how do we want to test it?

**Comment by sigmavirus24** [comment]
@pquentin I'd rather we seed/reseed the RNG used to shuffle so that we can log it and folks can maybe reproduce a bug. I'm to @jvanasco's point, this implies the need for a way to determine what IP actually was used to make a given request. 

I also agree that `socket_options` seems ... the wrong way to pass this in. I'm curious as to why that occurred to @sethmlarson though. I suspect there was good reason that you and I are missing

**Comment by sethmlarson** [comment]
@sigmavirus24 Mostly that if we already have a tunnel all the way down to that level of the library we could potentially reuse that tunnel since the shuffling of DNS addresses is socket-related. We're already using `**kwargs` everywhere so one more option shouldn't be a problem.

**Comment by pquentin** [comment]
@sigmavirus24 Can you please clarify what you believe is required for this feature to land? From what you said, it would be:

 * logging the chosen ip
 * accepting a RNG seed
 * logging the RNG seed

**Comment by sigmavirus24** [comment]
So I don't want to log the chosen ip, because we should attach the chosen IP programmatically to a response somehow (there's another recently active issue about this). 

If we aren't going to attach the information about the connection to a response or connection object somehow, we need to log the RNG seed that would allow the user to attempt to recreate the selection. I don't know if we need to accept the seed or allow them to replace a global `Random` instance with one seeded with the appropriate seed. This gives them the ability to debug if it's a consistent DNS result and maybe figure out what's happening there. I think users need significantly more visibility before we start introducing some amount of chaos.

**Comment by jvanasco** [comment]
@sigmavirus24 #1071 is for a 'debug' object for the ip; and a cross-reference to https://github.com/jakubroztocil/httpie/issues/549 for also tracking the active ssl certificate.  if this current ticket were to be implemented by attaching the IP to the response via a debug object/dict... I could most-likely follow up with a PR extending it to track the SSL Certificate.

**Comment by sigmavirus24** [comment]
Right. I'm not saying we need to do everything in that one issue to get to here. Just we should do what we need (minimally) from it to make this a piece of work that doesn't make users' lives worse. (Given how few API clients actually give users control over the underlying connection pieces)

**Comment by josephholsten** [comment]
Any consensus emerging? Are the requirements:

* keyword option to enable dns record shuffling
* accepting a RNG seed
* logging the RNG seed

And defer to other work to:
* log selected IP
